src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -140,57 +114,65 @@ uct_rdmacm_cm_query_reserved_qpn_cap_2(struct ibv_context *verbs,
         goto err_out;
     }
 
-    ctx->reserved_qpn_support                          = true;
-    ctx->reserved_qpn_ctx.log_reserved_qpn_granularity =
+    ctx->log_reserved_qpn_granularity =
             UCT_IB_MLX5DV_GET(cmd_hca_cap_2, cap, log_reserved_qpn_granularity);
     ucs_debug("reserved qpn cap: log_reserved_qpn_granularity is 0x%x",
-              ctx->reserved_qpn_ctx.log_reserved_qpn_granularity);
+              ctx->log_reserved_qpn_granularity);
+
+    ctx->reserved_qpn_support = 1;
+
+    ucs_spinlock_init(&ctx->lock, 0);
+    ucs_list_head_init(&ctx->blk_list);
     return UCS_OK;
 
 err_out:
 #endif
-    ctx->reserved_qpn_support = false;
-    return UCS_ERR_UNSUPPORTED;
+    ctx->reserved_qpn_support = 0;
+    /* Create a dummy completion queue */
+    ctx->cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
+    if (ctx->cq == NULL) {
+        ucs_error("ibv_create_cq() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdmacm_cm_reserved_qpn_blk_list_free(struct uct_rdmacm_cm_qp_context *ctx)
+static void
+uct_rdmacm_cm_device_context_cleanup(struct uct_rdmacm_cm_device_context *ctx)
 {
-    ucs_status_t status = UCS_OK;
-
 #if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    ucs_status_t status;
     struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
 
-    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+    ucs_spin_lock(&ctx->lock);
```

Comment:
why need to take lock, if we are destroying the context anyway and the lock will be destroyed as well?

Reply:
@yosefe Here we are deleting the block from block list and destroying qpn block from FW, so need lock to avoid multi-threads comptetition. The lock will be destoyed only after this function was called. I will move the lock destroying to the end of this function to make code more clear.

Reply:
@yosefe OK...I think the upper layer should make sure UCS_CLASS_CLEANUP_FUNC() will just be called only once, so there should be no competition here. Right?

Reply:
yes, the upper layer calls UCS_CLASS_CLEANUP_FUNC only once.